File: C_EXAMPLE/example_basic_AccessParameter.cpp
/**
 * @file example_basic_AccessParameter.cpp
 * @brief Example demonstrating how to access and modify basic L1 robot parameters.
 *
 * This example shows how to:
 * - Query the SDK and controller versions
 * - Read string, integer, and floating-point parameters from the robot system
 * - Modify a floating-point parameter and restore its original value
 *
 * @note This example is intended for users who need to understand the basic
 *       parameter read/write interfaces of the L1 robot SDK.
 */

#include "L1Robot.h"

/**
 * @brief Entry point of the example application.
 *
 * The workflow includes:
 * 1. Initialize communication with the robot controller
 * 2. Retrieve SDK and controller versions
 * 3. Read basic robot parameters
 * 4. Modify and verify an encoder error threshold parameter
 * 5. Restore the original parameter value
 *
 * @param[in] argc Argument count (unused)
 * @param[in] argv Argument vector (unused)
 * @return int Exit code (0 on normal exit)
 */
int main(int argc, char** argv)
{
    int sdk_version = 0;                 ///< SDK version number
    int controller_version = 0;          ///< Controller firmware version
    char robot_name[30] = { 0 };         ///< Robot name string
    int bus_freq = 0;                    ///< Communication bus frequency
    float new_enc_error_valve = 0.0f;    ///< Updated encoder error threshold
    float old_enc_error_valve = 0.0f;    ///< Original encoder error threshold

    /* Get SDK version */
    sdk_version = FX_L1_System_GetSDKVersion();
    printf("SDK version is 0x%08x\n", sdk_version);

    /* Establish communication with the controller */
    if (FX_L1_System_Link(6, 6, 7, 190, FX_LOG_ALL_FLAG) < 0)
    {
        printf("Failed to link system\n");
        goto WAIT_EXIT;
    }

    /* Get controller version */
    controller_version = FX_L1_System_GetControllerVersion();
    printf("Controller version is 0x%08x\n", controller_version);

    /* Read robot name */
    if (FX_L1_Param_GetString("R.BASIC.Name", robot_name) != FUNC_RET_SUCCESS)
    {
        printf("Failed to get string parameter: R.BASIC.Name\n");
    }

    /* Read bus frequency */
    if (FX_L1_Param_GetInt32("R.BASIC.BusFreq", &bus_freq) != 0)
    {
        printf("Failed to get int parameter: R.BASIC.BusFreq\n");
    }

    /* Read encoder error threshold */
    if (FX_L1_Param_GetFloat("R.A0.BASIC.EncErrorValve", &old_enc_error_valve) != FUNC_RET_SUCCESS)
    {
        printf("Failed to get float parameter: R.A0.BASIC.EncErrorValve\n");
    }

    /* Modify encoder error threshold */
    printf("Press any key to set float parameter R.A0.BASIC.EncErrorValve=2.5\n");
    getchar();
    if (FX_L1_Param_SetFloat("R.A0.BASIC.EncErrorValve", 2.5f) != FUNC_RET_SUCCESS)
    {
        printf("Failed to set float parameter R.A0.BASIC.EncErrorValve\n");
    }

    /* Verify modified parameter */
    printf("Press any key to get float parameter R.A0.BASIC.EncErrorValve\n");
    getchar();
    if (FX_L1_Param_GetFloat("R.A0.BASIC.EncErrorValve", &new_enc_error_valve) != FUNC_RET_SUCCESS)
    {
        printf("Failed to get float parameter: R.A0.BASIC.EncErrorValve\n");
    }

    /* Restore original parameter value */
    printf("Press any key to set float parameter R.A0.BASIC.EncErrorValve=%.4f\n", old_enc_error_valve);
    getchar();
    if (FX_L1_Param_SetFloat("R.A0.BASIC.EncErrorValve", old_enc_error_valve) != FUNC_RET_SUCCESS)
    {
        printf("Failed to set float parameter R.A0.BASIC.EncErrorValve\n");
    }

WAIT_EXIT:
    printf("Press any key to exit\n");
    getchar();
    return 0;
}

File: PYTHON_EXAMPLE/example_basic_AccessParameter.py
"""
@file example_basic_AccessParameter.py
@brief Example demonstrating how to access and modify basic L1 robot parameters.

Workflow overview:
    1. Initialize communication with the robot controller
    2. Retrieve SDK and controller versions
    3. Read basic robot parameters
    4. Modify and verify an encoder error threshold parameter
    5. Restore the original parameter value

@note This example is intended for users who need to understand the basic
       parameter read/write interfaces of the L1 robot SDK.
"""

import sys
import time
from pathlib import Path

root_dir = Path(__file__).parent.parent
if str(root_dir) not in sys.path:
    sys.path.insert(0, str(root_dir))
from PYTHON_SDK.GentoRobot import GentoRobot, FXLogMask, FXObjType,state_map

def main():

    log_mask=FXLogMask.FX_LOG_INFO_FLAG

    robot = GentoRobot()

    print(f"\n### 1/9 .Connecting...")
    ret = robot.link(6, 6, 7, 190, log_level=log_mask)
    if not robot._connected:
        print(f"Link failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return

    print(f"\n### 2/9. Getting versions...")
    print(f"Controller version:{robot.get_controller_version()}")
    print(f"Sdk version:{robot.get_sdk_version()}")

    print(f"\n### 3/9. Read robot name...")
    ret, value=robot.param_get_string("R.BASIC.Name")
    if ret!=0:
        print(f"Get parameter failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return
    print(f"R.BASIC.Name is: {value}")

    print(f"\n### 4/9. Read bus frequency...")
    ret,value=robot.param_get_int("R.BASIC.BusFreq")
    if ret!=0:
        print(f"Get parameter failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return
    print(f"R.BASIC.BusFreq is: {value}")

    print(f"\n### 5/9. Read encoder error threshold...")
    ret,cur_value=robot.param_get_float("R.A0.BASIC.EncErrorValve")
    if ret!=0:
        print(f"Get parameter failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return
    print(f"R.A0.BASIC.EncErrorValve is: {cur_value}")

    print(f"\n### 6/9. Modify encoder error threshold...")
    ret=robot.param_set_float("R.A0.BASIC.EncErrorValve", 2.5)
    if ret!=0:
        print(f"Set parameter failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return
    print(f'set "R.A0.BASIC.EncErrorValve" to 2.5')

    print(f"\n### 7/9. Verify modified parameter...")
    ret,value=robot.param_get_float("R.A0.BASIC.EncErrorValve")
    if ret!=0:
        print(f"Get parameter failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return
    print(f"R.A0.BASIC.EncErrorValve is: {value}")

    print(f"\n### 8/9. Restore original parameter value...")
    ret=robot.param_set_float("R.A0.BASIC.EncErrorValve",cur_value)
    if ret!=0:
        print(f"Set parameter failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return
    print(f'set "R.A0.BASIC.EncErrorValve" to {cur_value}')

    print(f"\n### 9/9. Task finished.")

if __name__ == "__main__":
    main()
